### PR TITLE
AST-38524 Vorpal extension

### PR DESCRIPTION
**Pull request by tamarleviCm**
### Description

> Vorpal is a new light engine that runs in costumer's computer
the engine scans the files as the developer code on them
 

### References

> https://checkmarx.atlassian.net/browse/AST-38524

### Testing

> Automation Tests for the new checkbox in the vscode settings
>
> manual tests for e2e, parallel, and an edge cases

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR (if applicable).
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used

### CODE REVIEW

File: src/commands/vorpalCommand.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  public registerVorpal() {
+    const vorpalActive = vscode.workspace
+      .getConfiguration("CheckmarxVorpal")
+      .get("Activate Vorpal Auto Scanning") as boolean;
+    if (vorpalActive) {
+      this.installVorpal();
+      this.registerVorpalScanOnChangeText();
+    } else {
+      this.disposeVorpalScanOnChangeText();
+      //remove Vorpal problems
```
Comment (OrShamirCM) [inline src/commands/vorpalCommand.ts:23]: the method name is enough in this case :)

File: src/cx/cx.ts [hunk 2]
```diff
@@ -421,4 +420,38 @@ export class Cx implements CxPlatform {
...
+	
+	  async scanVorpal(sourcePath: string): Promise<CxVorpal> {
+		let config = this.getAstConfiguration();
+		if (!config) {
+		  config = new CxConfig();
+		}
+		const cx = new CxWrapper(config);
+		const scans = await cx.scanVorpal(sourcePath, false, constants.vsCodeAgent);
+		if (scans.payload && scans.exitCode === 0) {
+		  return scans.payload[0];
```
Comment (OrShamirCM) [inline src/cx/cx.ts:448]: this check may be in private method as we use it multiple time

File: src/vorpal/scanVorpal.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import CxVorpal from "@checkmarxdev/ast-cli-javascript-wrapper/dist/main/vorpal/CxVorpal";
+import { constants } from "../utils/common/constants";
+
+const vorpalDir = "CxVorpal";
+
+export const diagnosticCollection = vscode.languages.createDiagnosticCollection(
+  constants.extensionFullName
+);
+
+export async function scanVorpal(document: vscode.TextDocument, logs: Logs) {
```
Comment (OrShamirCM) [inline src/vorpal/scanVorpal.ts:17]: class name can be maybe chang to vorpalService

File: src/vorpal/scanVorpal.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  } catch (error) {
+    console.error(error);
+    logs.error(constants.errorScanVorpal);
+  }
+}
+
+function ignoreFiles(document: vscode.TextDocument): boolean {
+  // ignore our output log file, settings.json
+  if (document.languageId.toUpperCase() == 'LOG' ||
+    (path.basename(document.uri.fsPath) === 'settings.json' && document.uri.fsPath.includes("User"))) {
```
Comment (OrShamirCM) [inline src/vorpal/scanVorpal.ts:57]: suggestion -it can be nice to pass here a list that we are excluding

File: src/vorpal/scanVorpal.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+function updateProblems(scanVprpalResult: CxVorpal, uri: vscode.Uri) {
+  diagnosticCollection.delete(uri);
+  const diagnostics: vscode.Diagnostic[] = [];
+
+  for (let i = 0; i < scanVprpalResult.scanDetails.length; i++) {
+    const res = scanVprpalResult.scanDetails[i];
+    const range = new vscode.Range(
+      new vscode.Position(res.line - 1, 0),
```
Comment (OrShamirCM) [inline src/vorpal/scanVorpal.ts:74]: what is this range?

File: src/vorpal/scanVorpal.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    );
+    diagnostic.source = constants.vorpalEngineName;
+    diagnostics.push(diagnostic);
+  }
+  diagnosticCollection.set(uri, diagnostics);
+}
+
+function parseSavirity(vorpalSavirity: string) {
+  switch (vorpalSavirity.toUpperCase()) {
+    case "HIGH":
```
Comment (OrShamirCM) [inline src/vorpal/scanVorpal.ts:90]: should we add critical?

**Review by OrShamirCM — COMMENTED** [review]
Great work!

File: src/cx/cx.ts [hunk 7]
```diff
@@ -421,4 +420,38 @@ export class Cx implements CxPlatform {
...
 		show ? statusBarItem.show() : statusBarItem.hide();
 	}
-}
+	async installVorpal(): Promise<CxVorpal> {
+		let config = this.getAstConfiguration();
+		if (!config) {
+		  config = new CxConfig();
+		}
+		const cx = new CxWrapper(config);
+		const scans = await cx.scanVorpal(null, true, constants.vsCodeAgent);
```
[the remark this answers is not in the document]
Comment (tamarleviCm) [reply src/cx/cx.ts:433]: It is perceived higher

File: src/utils/common/commands.ts [hunk 8]
```diff
@@ -102,6 +102,9 @@ export const commands = {
   refreshScaTree: `${constants.extensionName}.refreshSCATree`,
   clearSca: `${constants.extensionName}.clearSca`,
 
+  installVorpal: `${constants.extensionName}.installVorpal`,
```
Comment (tamarleviCm) [inline src/utils/common/commands.ts:105]: remove

File: src/vorpal/scanVorpal.ts (hunk 5 above)
Comment (tamarleviCm) [reply src/vorpal/scanVorpal.ts:70]: Since we don't get the location of the problem - only the row, I hardcoded the location from 0-100.

File: src/extension.ts [hunk 9]
```diff
@@ -23,11 +23,13 @@ import { WorkspaceListener } from "./utils/listener/workspaceListener";
 import { DocAndFeedbackView } from "./views/docsAndFeedbackView/docAndFeedbackView";
 import { messages } from "./utils/common/messages";
 import { commands } from "./utils/common/commands";
+import { VorpalCommand } from "./commands/vorpalCommand";
 
 export async function activate(context: vscode.ExtensionContext) {
   // Create logs channel and make it visible
   const output = vscode.window.createOutputChannel(constants.extensionFullName);
   const logs = new Logs(output);
+  logs.show();
```
Comment (AlvoBen) [inline src/extension.ts:32]: If i'm not wrong, we add this logs.show() in other place, there was a story created to remove it because it's annoying some of the customers: https://checkmarx.atlassian.net/browse/AST-37522

File: src/test/1.settings.test.ts [hunk 10]
```diff
@@ -60,4 +60,25 @@ describe("Extension settings tests", () => {
     const enablement = await setting.getValue();
     expect(enablement).to.equal(true);
   });
+
+  it("verify vorpal checkbox exists in the settings", async function () {
+    settingsEditor = await bench.openSettings();
+    const vorpalCheckbox = await settingsEditor.findSetting(
+      "ActivateVorpalAutoScanning",
+      "Checkmarx Vorpal"
```
Comment (AlvoBen) [inline src/test/1.settings.test.ts:68]: maybe make "ActivateVorpalAutoScanning", "Checkmarx Vorpal" as constants? I saw that those values are also being use in other places

File: src/test/9.vorpal.test.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
+describe("Vorpal engine tests", () => {
+
+  before(async function () {
+  });
+ 
+  after(async () => {
+  });
+ 
+  // it("verify vorpal checkbox exists in the settings", async function () {}); // moved to 1.settings.test
```
Comment (AlvoBen) [inline src/test/9.vorpal.test.ts:9]: should be removed?

File: src/vorpal/scanVorpal.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+  for (let i = 0; i < scanVorpalResult.scanDetails.length; i++) {
+    const res = scanVorpalResult.scanDetails[i];
+    const range = new vscode.Range(
+      new vscode.Position(res.line - 1, 0),
+      new vscode.Position(res.line - 1, 100)
+    );
+    const diagnostic: vscode.Diagnostic = new vscode.Diagnostic(
+      range,
+      res.ruleName + " - " + res.remediationAdvise,
```
Comment (AlvoBen) [inline src/vorpal/scanVorpal.ts:75]: suggestion: \`${res.ruleName} - ${res.remediationAdvise}\`

File: src/extension.ts (hunk 9 above)
Comment (tamarleviCm) [reply src/extension.ts:32]: It's interesting that it is not merged into my PR...
I will fix it

File: src/cx/cx.ts [hunk 13]
```diff
@@ -421,4 +420,38 @@ export class Cx implements CxPlatform {
...
+		  const errorRes = new CxVorpal();
+		  errorRes.error = scans.status;
+		  return errorRes;
+		}
+	  }
+	
+	  async scanVorpal(sourcePath: string): Promise<CxVorpal> {
+		let config = this.getAstConfiguration();
+		if (!config) {
+		  config = new CxConfig();
```
[the remark this answers is not in the document]
Comment (tamarleviCm) [reply src/cx/cx.ts:451]: right
but we will get the intuitive error from the cli and will show it to the user
in this way, we don't depend on the cli logic.

**Review by OrShamirCM — APPROVED** [review]
